src/neml2/models/Model.cxx

```diff
@@ -32,45 +32,254 @@ ParameterSet
+          dout_dp.index_put_({b, i, torch::indexing::Slice()}, jac_row);
+        }
+        catch (c10::Error &)
+        {
+          // This is aggravating: libTorch throws if the derivative is zero... but why?!
+        }
+      }
+  }
+  else
+  {
+    for (TorchSize i = 0; i < outval.base_sizes()[0]; i++)
+    {
+      BatchTensor<1> grad_outputs = torch::zeros_like(outval);
+      grad_outputs.index_put_({torch::indexing::Ellipsis, i}, 1.0);
+      try
+      {
+        auto jac_row = torch::autograd::grad({outval}, {pval}, {grad_outputs}, true)[0];
+        dout_dp.base_index_put({i, torch::indexing::Slice()}, jac_row);
+      }
+      catch (c10::Error &)
+      {
+        // This is aggravating: libTorch throws if the derivative is zero... but why?!
+      }
+    }
+  }
+
+  return dout_dp;
+}
+
 LabeledVector
-Model::value(LabeledVector in) const
+Model::value(const LabeledVector & in) const
 {
-  LabeledVector out(in.batch_size(), output());
-  set_value(in, out);
+  LabeledVector out(in.batch_size(), {&output()}, in.options());
+  set_value(in, &out);
   return out;
 }
 
 LabeledMatrix
-Model::dvalue(LabeledVector in) const
+Model::dvalue(const LabeledVector & in) const
 {
-  auto [out, dout_din] = value_and_dvalue(in);
+  if (_AD_1st_deriv)
+    return std::get<1>(value_and_dvalue(in));
+
+  LabeledMatrix dout_din(in.batch_size(), {&output(), &in.axis()}, in.options());
+  set_value(in, nullptr, &dout_din);
   return dout_din;
 }
 
+LabeledTensor3D
+Model::d2value(const LabeledVector & in) const
+{
+  if (_AD_2nd_deriv)
+    return std::get<2>(value_and_dvalue_and_d2value(in));
+
+  LabeledTensor3D d2out_din2(in.batch_size(), {&output(), &in.axis(), &in.axis()}, in.options());
+  set_value(in, nullptr, nullptr, &d2out_din2);
+  return d2out_din2;
+}
+
 std::tuple<LabeledVector, LabeledMatrix>
-Model::value_and_dvalue(LabeledVector in) const
+Model::value_and_dvalue(const LabeledVector & in) const
 {
-  LabeledVector out(in.batch_size(), output());
-  LabeledMatrix dout_din(out, in);
-  set_value(in, out, &dout_din);
+  LabeledVector out(in.batch_size(), {&output()}, in.options());
+  LabeledMatrix dout_din(in.batch_size(), {&output(), &in.axis()}, in.options());
+
+  if (_AD_1st_deriv)
+  {
+    // Set requires_grad to true if not already
+    bool req_grad = in.tensor().requires_grad();
+    in.tensor().requires_grad_();
+    // Evaluate the model value
+    set_value(in, &out);
+    // Loop over rows to retrieve the derivatives
+    for (TorchSize i = 0; i < out.tensor().base_sizes()[0]; i++)
+    {
+      BatchTensor<1> grad_outputs = torch::zeros_like(out.tensor());
+      grad_outputs.index_put_({torch::indexing::Ellipsis, i}, 1.0);
+      try
+      {
+        auto jac_row =
+            torch::autograd::grad({out.tensor()}, {in.tensor()}, {grad_outputs}, true, true)[0];
+        dout_din.tensor().base_index_put({i, torch::indexing::Slice()}, jac_row);
+      }
+      catch (c10::Error &)
```

Comment:
Fix this either by:

1. Figuring out the right torch options to grad...
2. Detecting the issue with logic in the loop versus catching here
3. Worst case: check error string...

Reply:
Fixed -- see above.